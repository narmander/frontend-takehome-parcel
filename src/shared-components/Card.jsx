import React from 'react';
import styled, { css } from 'styled-components';
import PropTypes from 'prop-types';

import { Ruby } from '../assets/icons/Ruby';
import { RED, GRAY, OFF_WHITE } from '../assets/colors';

// component is taking in custom props but could easily be switched to be more generic 
// and take different props when scaling for reuse.

export const Card = ({
  downloads,
  href,
  info,
  onIconClick,
  title,
  version,
  toggled,
  ...props
}) => {
  return (
    <CardStyled {...props}>
      <Icon onClick={onIconClick} saved={toggled} />
      <div style={{ flexGrow: 1 }}>
        <div>
          <Title>{title}</Title>
          <SubTitle>{version}</SubTitle>
        </div>
        <Description>{info}</Description>
      </div>
      <Details>
        {downloads}
        <Label>downloads</Label>
        <InfoLink href={href} target="blank">
          More Info
        </InfoLink>
      </Details>
    </CardStyled>
  );
};

const CardStyled = styled.div`
  padding: 15px;
  position: relative;
  background-color: ${OFF_WHITE};
  box-shadow: 0 0 8px 4px rgba(0, 0, 0, 0.1);
  border-radius: 5px;
  transition: ease-in-out 120ms;
  display: flex;
  &:hover {
    transform: scale(1.015);
  }

  width: 100%;
  margin: 30px;

  /* cell phone */
  @media screen and (min-width: 400px) {
    width: 60%;
    margin: 32px auto;
  }

  /* tablet */
  @media screen and (min-width: 700px) {
    width: calc(50% - 64px);
  }

  /* laptop/desktop */
  @media screen and (min-width: 1000px) {
    width: calc(50% - 64px);
  }
`;

// const Icon = styled(({saved, ...props}) => <Ruby {...props} />)`
const Icon = styled(Ruby)`
  width: 50px;
  height: 50px;
  min-width: 50px;
  max-width: 50px;
  cursor: pointer;
  transition: ease-in-out 120ms;

  path {
    stroke: ${RED};
    transition: ease-in-out 120ms;
  }

  &:hover {
    transform: translateY(-2px) scale(1.2);
    path {
      fill: rgba(255, 0, 0, 0.5);
      stroke: white;
    }
  }

  ${props =>
    props.saved &&
    css`
      path {
        fill: ${RED};
        stroke: white;
      }

      &:hover {
        path {
          fill: ${RED};
        }
      }
    `}
`;

const Title = styled.h3`
  margin: 0 16px 0 0;
  font-size: 24px;
  font-weight: 700;
  display: inline-block;
`;

const SubTitle = styled.span`
  display: inline-block;
`;

const Description = styled.p`
  font-family: Lato, sans-serif;
  color: ${GRAY};
  margin-right: 5px;
`;

const Details = styled.div`
  text-align: right;
  width: 70px;
  position: relative;
  margin-top: 3px;
`;

const Label = styled.span`
  display: block;
  color: ${GRAY};
  font-family: Lato, sans-serif;
`;

const InfoLink = styled.a`
  position: absolute;
  display: block;
  text-decoration: none;
  font-family: Lato, sans-serif;
  transition: ease-in-out 120ms;
  width: 70px;

  &:link {
    color: ${RED};
  }

  &:hover {
    transform: translateY(-1px) scale(1.1);
    color: ${RED};
  }
`;


Card.PropTypes = {
  downloads: PropTypes.number,
  href: PropTypes.string,
  info: PropTypes.string,
  onIconClick: PropTypes.func,
  title: PropTypes.string,
  version: PropTypes.string,
  toggled: PropTypes.func,
}