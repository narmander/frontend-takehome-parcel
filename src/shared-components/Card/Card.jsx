import React, { useState, useEffect } from 'react';
import styled, { css } from 'styled-components';

import { Ruby as R } from '../../assets/icons/Ruby';

export const Card = ({
  downloads,
  href,
  info,
  onIconClick,
  title,
  version,
  toggled,
  ...props
}) => {
  return (
    <CardStyled {...props}>
      <Ruby onClick={onIconClick} saved={toggled ? 1 : 0}/>
      <div style={{ flexGrow: 1 }}>
        <div>
          <Title>{title}</Title>
          <Version>{version}</Version>
        </div>
        <Description>{info}</Description>
      </div>
      <Downloads>
        {downloads}
        <Label>downloads</Label>
      </Downloads>
    </CardStyled>
  );
};

const Label = styled.span`
  display: block;
  font-family: Lato, sans-serif;
`;

const CardStyled = styled.div`
  padding: 16px;
  box-shadow: 0 0 8px 4px rgba(0, 0, 0, 0.1);
  border-radius: 5px;
  transition: ease-in-out 120ms;
  display: flex;
  &:hover {
    transform: scale(1.015);
  }

  width: 100%;
  margin: 32px;

  /* tablet */
  @media screen and (min-width: 640px) {
    /* width: calc(50% - 64px); */
    width: 60%;
    margin: 32px auto;
  }

  /* laptop */
  @media screen and (min-width: 800px) {
    width: calc(50% - 64px);
  }

  /* desktop */
  @media screen and (min-width: 1000px) {
    width: calc(33% - 64px);
  }
`;

const Ruby = styled(R)`
  width: 50px;
  height: 50px;
  min-width: 50px;
  max-width: 50px;
  cursor: pointer;
  transition: ease-in-out 120ms;

  path {
    stroke: red;
    transition: ease-in-out 120ms;
  }

  ${props =>
    !!props.saved &&
    css`
      path {
        fill: red;
        stroke: white;
      }
    `}

  &:hover {
    transform: translateY(-2px) scale(1.2);
    path {
      fill: red;
      stroke: white;
    }
  }
`;

const Title = styled.h3`
  margin: 0 16px 0 0;
  font-size: 24px;
  font-weight: 700;
  display: inline-block;
`;

const Version = styled.span`
  display: inline-block;
`;

const Description = styled.p`
  font-family: Lato, sans-serif;
`;

const Downloads = styled.span`
  text-align: right;
  width: 150px;
`;
