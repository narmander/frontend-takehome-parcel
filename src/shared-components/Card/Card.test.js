import React from 'react';
import renderer from 'react-test-renderer';

import { Card } from './Card.jsx';

test('Icon turns red when clicked', () => {

    const props = {toggled: false}

  const component = renderer.create(
    <Card {...props}/>
  );

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
