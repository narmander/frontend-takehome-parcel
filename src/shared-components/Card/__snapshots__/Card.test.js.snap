// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon turns red when clicked 1`] = `
<div
  className="sc-bdVaJa jnBuUC"
>
  <svg
    className="sc-bwzfXH gGkBbC"
    saved={0}
    viewBox="0 0 64 80"
  >
    <g>
      <path
        d="m 32,52.000023 -20,-25 10,-15 20,0 10,15 z"
        fill="none"
        stroke="BLACK"
        strokeDasharray="none"
        strokeLinecap="butt"
        strokeLinejoin="round"
        strokeMiterlimit="4"
        strokeOpacity="1"
        strokeWidth="1.5"
      />
      <path
        d="m 37,12.000023 5,15 -10,25"
        fill="none"
        stroke="BLACK"
        strokeDasharray="none"
        strokeLinecap="butt"
        strokeLinejoin="round"
        strokeMiterlimit="4"
        strokeOpacity="1"
        strokeWidth="1.5"
      />
      <path
        d="m 27,12.000023 -5,15 10,25"
        fill="none"
        stroke="BLACK"
        strokeDasharray="none"
        strokeLinecap="butt"
        strokeLinejoin="round"
        strokeMiterlimit="4"
        strokeOpacity="1"
        strokeWidth="1.5"
      />
      <path
        d="m 52,27.000023 -40,0"
        fill="none"
        stroke="BLACK"
        strokeDasharray="none"
        strokeLinecap="butt"
        strokeLinejoin="round"
        strokeMiterlimit="4"
        strokeOpacity="1"
        strokeWidth="1.5"
      />
    </g>
  </svg>
  <div
    style={
      Object {
        "flexGrow": 1,
      }
    }
  >
    <div>
      <h3
        className="sc-htpNat iSsIGQ"
      />
      <span
        className="sc-bxivhb jzkpli"
      />
    </div>
    <p
      className="sc-ifAKCX flkLYl"
    />
  </div>
  <div
    className="sc-EHOje igtscQ"
  >
    <span
      className="sc-bZQynM opGmY"
    >
      downloads
    </span>
    <a
      className="sc-gzVnrw gVXohH"
      target="blank"
    >
      More Info
    </a>
  </div>
</div>
`;
